libc/gcc/Makefile

```diff
@@ -2,7 +2,10 @@
 
 include $(TOPDIR)/libc/Makefile.inc
 
-SRCS = divmodsi3.s ldivmod.s ashlsi3.s
+# It is feasible to now link programs with the real libgcc, so this libgcc
+# module is not really needed any more.  -- tkchia 20190603
+# SRCS = divmodsi3.s ldivmod.s ashlsi3.s
+SRCS =
```

Comment:
Hello @mfld-fr,

The current `libgcc` implementations of these routines are [here](https://github.com/tkchia/gcc-ia16/blob/21b95e80adb7a080b850a9a047505ebc653ca580/libgcc/config/ia16/lib1funcs.S).  They are also coded in assembly language.  The `__ashlsi3` implementation is a little bit larger, but probably faster, than the version currently in the ELKS code.  I have not really closely compared the division routines yet.

Thank you!

Reply:
Hello @mfld-fr,

Interestingly, the `libgcc` core routine for 32-bit division (`__ia16_ldivmodu`, taken from the FreeDOS kernel sources) is `0x4e` bytes --- less than half the size (!) of the `elks-libc` core routine (`ldivmod` + `ludivmod`), which is `0xda` bytes.  However, the `libgcc` implementation does not handle signed division yet.

I will try to implement signed 32-bit division in `libgcc` in assembly, and see how things go.

Thank you!

Reply:
Hello @mfld-fr,

I have incorporated your `__ashlsi3` implementation into the `libgcc` code in the `-Os` case (see https://github.com/tkchia/gcc-ia16/commit/2dd73a5500dce4be8f0f7d4bab02f5ed6bf849b0), and further squished it --- so that it also limits the total shift count to `0x1f` but still takes up the same number of bytes.  (It is probably a bit slower now than the `elks-libc` implementation, though.)
```
00000000 <__ashlsi3>:
   0:	5b                   	pop    %bx
   1:	58                   	pop    %ax
   2:	5a                   	pop    %dx
   3:	59                   	pop    %cx
   4:	51                   	push   %cx
   5:	52                   	push   %dx
   6:	50                   	push   %ax
   7:	53                   	push   %bx
   8:	83 e1 1f             	and    $0x1f,%cx
   b:	e3 06                	jcxz   13 <__ashlsi3+0x13>
   d:	d1 e0                	shl    %ax
   f:	d1 d2                	rcl    %dx
  11:	e2 fa                	loop   d <__ashlsi3+0xd>
  13:	c3                   	ret    
```

I have also added signed 32-bit division routines `__divsi3` and `__modsi3`, based on the existing unsigned division routine taken earlier from the FreeDOS kernel code.  The new implementation is shorter, and --- I think --- more conformant to C11, since it rounds quotients towards 0, as mandated by the standard.  (The `elks-libc` implementation in contrast rounds quotients towards −∞.)

I am still testing the changes and will let you know once this is done.

Thank you!

Reply:
I have completed the tests and bumped up the GCC-IA16 version number.